lib/logstash/inputs/kafka.rb

```diff
@@ -20,168 +21,89 @@
+  config :session_timeout_ms, :validate => :string, :default => "30000"
+  config :value_deserializer, :validate => :string, :default => "org.apache.kafka.common.serialization.StringDeserializer"
+  config :num_threads, :validate => :number, :default => 1
+  config :topics, :validate => :array, :required => true
+  config :timeout_ms, :validate => :number, :default => 100
+  config :max_messages, :validate => :number
 
   public
   def register
-    LogStash::Logger.setup_log4j(@logger)
-    options = {
-        :zk_connect => @zk_connect,
-        :group_id => @group_id,
-        :topic_id => @topic_id,
-        :auto_offset_reset => @auto_offset_reset,
-        :rebalance_max_retries => @rebalance_max_retries,
-        :rebalance_backoff_ms => @rebalance_backoff_ms,
-        :consumer_timeout_ms => @consumer_timeout_ms,
-        :consumer_restart_on_error => @consumer_restart_on_error,
-        :consumer_restart_sleep_ms => @consumer_restart_sleep_ms,
-        :consumer_id => @consumer_id,
-        :fetch_message_max_bytes => @fetch_message_max_bytes,
-        :allow_topics => @white_list,
-        :filter_topics => @black_list,
-        :value_decoder_class => @decoder_class,
-        :key_decoder_class => @key_decoder_class
-    }
-    if @reset_beginning
-      options[:reset_beginning] = 'from-beginning'
-    end # if :reset_beginning
-    topic_or_filter = [@topic_id, @white_list, @black_list].compact
-    if topic_or_filter.count == 0
-      raise LogStash::ConfigurationError, 'topic_id, white_list or black_list required.'
-    elsif topic_or_filter.count > 1
-      raise LogStash::ConfigurationError, 'Invalid combination of topic_id, white_list or black_list. Use only one.'
-    end
-    @kafka_client_queue = SizedQueue.new(@queue_size)
-    @consumer_group = create_consumer_group(options)
-    @logger.info('Registering kafka', :group_id => @group_id, :topic_id => @topic_id, :zk_connect => @zk_connect)
+    @message_count = 0
+    @runner_threads = []
   end # def register
 
   public
   def run(logstash_queue)
-    # noinspection JRubyStringImportInspection
-    java_import 'kafka.common.ConsumerRebalanceFailedException'
-    @logger.info('Running kafka', :group_id => @group_id, :topic_id => @topic_id, :zk_connect => @zk_connect)
-    begin
-      @consumer_group.run(@consumer_threads,@kafka_client_queue)
-
-      while !stop?
-        if !@kafka_client_queue.empty?
-          event = @kafka_client_queue.pop
-          queue_event(event, logstash_queue)
-        end
-      end
-
-      until @kafka_client_queue.empty?
-        queue_event(@kafka_client_queue.pop,logstash_queue)
-      end
-
-      @logger.info('Done running kafka input')
-    rescue => e
-      @logger.warn('kafka client threw exception, restarting',
-                   :exception => e)
-      Stud.stoppable_sleep(Float(@consumer_restart_sleep_ms) * 1 / 1000) { stop? }
-      retry if !stop?
-    end
+    @runner_consumers = num_threads.times.map { || new_consumer }
```

Comment:
From Kafka's documentation - http://people.apache.org/~nehanarkhede/kafka-0.9-consumer-javadoc/doc/ 

> The consumer is thread safe and should generally be shared among all threads for best performance.

Looks like we just need one consumer instance and share it across threads...